lib/src/chunking.rs

```diff
@@ -364,78 +400,369 @@ impl Chunking {
-/// - Take the whole tail and group them toether (this is the overly simplistic part)
-fn basic_packing(components: &[ObjectSourceMetaSized], bins: NonZeroU32) -> Vec<ChunkedComponents> {
-    // let total_size: u64 = components.iter().map(|v| v.size).sum();
-    // let avg_size: u64 = total_size / components.len() as u64;
+
+// Total available bins = n
+//
+// 1 bin for all the u32_max frequency pkgs
+// 1 bin for all newly added pkgs
+// 1 bin for all low size pkgs
+//
+// 60% of n-3 bins for high size pkgs
+// 40% of n-3 bins for medium size pkgs
+//
+// If HS bins > limit, spillover to MS to package
+// If MS bins > limit, fold by merging 2 bins from the end
+//
+fn basic_packing<'a>(
+    components: &'a [ObjectSourceMetaSized],
+    bin_size: NonZeroU32,
+    prior_build_metadata: Option<oci_spec::image::ImageManifest>,
+) -> Vec<Vec<&'a ObjectSourceMetaSized>> {
     let mut r = Vec::new();
-    // And handle the easy case of enough bins for all components
-    // TODO: Possibly try to split off large files?
-    if components.len() <= bins.get() as usize {
-        r.extend(components.iter().map(|v| vec![v]));
-        return r;
-    }
-    // Create a mutable copy
     let mut components: Vec<_> = components.iter().collect();
-    // Iterate over the component tail, folding by source id
-    let mut by_src = HashMap::<_, Vec<&ObjectSourceMetaSized>>::new();
-    // Take the tail off components, then build up mapping from srcid -> Vec<component>
-    for component in components.split_off(bins.get() as usize) {
-        by_src
-            .entry(&component.meta.srcid)
-            .or_default()
-            .push(component);
+    let before_processing_pkgs_len = components.len();
+
+    //If the current rpm-ostree commit to be encapsulated is not the one in which packing structure changes, then
+    //  Flatten out prior_build_metadata to view all the packages in prior build as a single vec
+    //  Compare the flattened vector to components to see if pkgs added, updated,
+    //  removed or kept same
+    //  if pkgs added, then add them to the last bin of prior
+    //  if pkgs removed, then remove them from the prior[i]
+    //  iterate through prior[i] and make bins according to the name in nevra of pkgs to update
+    //  required packages
+    //else if pkg structure to be changed || prior build not specified
+    //  Recompute optimal packaging strcuture (Compute partitions, place packages and optimize build)
+
+    if let Some(prior_build) = prior_build_metadata {
+        tracing::debug!("Keeping old package structure");
+
+        //1st layer is skipped as packing doesn't manage ostree_commit layer
+        let mut curr_build: Vec<Vec<String>> = prior_build
+            .layers()
+            .iter()
+            .skip(1)
+            .map(|layer| {
+                let annotation_layer = layer
+                    .annotations()
+                    .as_ref()
+                    .expect("Layer does not consist annotation");
```

Comment:
I think instead of panic'ing an explicit error would probably be better here.  This would come up in a real world scenario I think where a build system passes a prior build that was generated *before* this code landed.

So something more like:

```suggestion
                let annotation_pkgs = layer
                    .annotations()
                    .as_ref()
                    .and_then(|annos|  annos.get(CONTENT_ANNOTATION))
                    .ok_or_else(|| anyhow!("Prior build is missing annotation {CONTENT_ANNOTATION}");
```